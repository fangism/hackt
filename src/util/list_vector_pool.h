/**
	\file "list_vector_pool.h"
	Simple template container-based memory pool.  
	Basically allocates a large chunk at a time.  

	$Id: list_vector_pool.h,v 1.1 2004/11/05 02:38:47 fang Exp $
 */

#ifndef	__LIST_VECTOR_POOL_H__
#define	__LIST_VECTOR_POOL_H__

#include <pthread.h>

#include <assert.h>
#include <queue>
#include <list>
#include <vector>


// because FreeBSD's <pthreads.h> defines initializer as NULL
// and I haven't figured out a workaround yet.
#if defined(__FreeBSD__)
// #warn	"This allocator will not be thread-safe in FreeBSD... yet."
#define	THREADED_ALLOC		0
#else
#define	THREADED_ALLOC		1
#endif

// turn invariant assertions on or off
#define	DEBUG_LIST_VECTOR_POOL	1

// annoying debug messages
#define VERBOSE_ALLOC		1 && DEBUG_LIST_VECTOR_POOL
#define	VERBOSE_LOCK		0 && VERBOSE_ALLOC

#if VERBOSE_ALLOC
#include <iostream>
#endif

#if DEBUG_LIST_VECTOR_POOL
#define INVARIANT_ASSERT(foo)	assert(foo)
#else
#define INVARIANT_ASSERT(foo)	{ }
#endif

//=============================================================================
/**
	Memory-management utility namespace.
	Should implement the std::allocator interface.  
 */
namespace memory {
typedef	pthread_mutex_t				mutex_type;
using std::queue;
using std::list;
using std::vector;

#if VERBOSE_ALLOC
using namespace std;		// for debugging
#endif

//=============================================================================
// forward declaration
template <class T, bool Threaded = true>
class list_vector_pool;

//-----------------------------------------------------------------------------
// specialization
template <>
class list_vector_pool<void> {
public:
	typedef size_t		size_type;
	typedef ptrdiff_t	difference_type;
	typedef void*		pointer;
	typedef const void*	const_pointer;
	typedef void		value_type;
// no rebind yet...
};

//-----------------------------------------------------------------------------
#if 0
// specialization for when allocated type is just pointer, 
// to prevent pointer recursion, not needed unless free_list_type is fancy
template <class T, bool Threaded>
class list_vector_pool<T*, Threaded> {
public:
	typedef	T*				value_type;
	typedef	size_t				size_type;
	typedef	ptrdiff_t			difference_type;
	typedef	value_type*			pointer;
	typedef	const value_type*		const_pointer;
	typedef	value_type&			reference;
	typedef	const value_type&		const_reference;
private:
	typedef	vector<value_type>		chunk_type;
	typedef	list<chunk_type>		impl_type;
	// plain queue to stop pointer recursion
	typedef	queue<pointer>			free_list_type;
public:

};	// end class list_vector_pool specialization
#endif

//-----------------------------------------------------------------------------
/**
	Simple, efficient memory pool.  
	No alignment or placement.  
	Only good for allocating one element at a time, 
	so is not general purpose.  
	For a more general pool allocator use the __gnu_cxx::__pool_alloc, 
	which is likely to be the default allocator.  
 */
template <class T, bool Threaded>
class list_vector_pool {
public:
	typedef	T				value_type;
	typedef	size_t				size_type;
	typedef	ptrdiff_t			difference_type;
	typedef	T*				pointer;
	typedef	const T*			const_pointer;
	typedef	T&				reference;
	typedef	const T&			const_reference;
private:
	// could propagate thread-sensitivity parameter to the respective
	// STL data-structures' allocators...
	typedef	vector<value_type>		chunk_type;
	typedef	list<chunk_type>		impl_type;
#if 1
	typedef	queue<pointer>			free_list_type;
#else
	// ballsy: use itself as allocator! (template recursion problem?)
	typedef	queue<pointer, deque<pointer, list_vector_pool<pointer> > >
						free_list_type;
#endif
#if THREADED_ALLOC
	static mutex_type			the_mutex;
	friend struct Lock;			// grant access to the_mutex
	struct Lock {
		Lock() {
//			assert(the_mutex);
#if VERBOSE_LOCK
			cerr << "getting lock at " << &the_mutex << "... ";
#endif
			if (Threaded)
				assert(!pthread_mutex_lock(&the_mutex));
#if VERBOSE_LOCK
			cerr << "got it." << endl;
#endif
		}

		~Lock() {
#if VERBOSE_LOCK
			cerr << "releasing lock... ";
#endif
			if (Threaded)
				assert(!pthread_mutex_unlock(&the_mutex));
#if VERBOSE_LOCK
			cerr << "released." << endl;
#endif
		}
	};	// end struct Lock
	// may need __attribute__ ((__unused__))
#endif	// THREADED_ALLOC

private:
	// probably no need to change...
	const size_t				chunk_size;
	impl_type				pool;
	free_list_type				free_list;
public:
	/**
		Reserves one chunk up-front.  
		\param C is number of elements to allocate at a time.
			Should be related to page_size/sizeof(T).  
	 */
	list_vector_pool(const size_type C = 16) : 
			chunk_size(C), pool(), free_list() {
		assert(chunk_size);
		// worry about alignment, placement and pages sizes later
		pool.push_back(chunk_type());
		pool.back().reserve(chunk_size);
		INVARIANT_ASSERT(pool.back().capacity() == chunk_size);
		INVARIANT_ASSERT(!pool.back().size());
		INVARIANT_ASSERT(free_list.empty());
	}

	// no copy-constructor

	// default destructor suffices

	/// allocate one element only, without construction
	pointer
	allocate(void) {
#if THREADED_ALLOC
		// volatile? do not optimize away?
		Lock got_the_mutex;
#endif
		// see if an address is available from the free-list
		if (!free_list.empty()) {
			pointer ret = free_list.front();
			INVARIANT_ASSERT(ret);
			free_list.pop();
#if VERBOSE_ALLOC
			cerr << "alloc from free-list @ " << ret << endl;
#endif
			return ret;
		} else {
			// if not, use the chunk.
			INVARIANT_ASSERT(!pool.empty());
			chunk_type* chunk = &pool.back();
			INVARIANT_ASSERT(chunk->capacity());
			if (chunk->size() == chunk->capacity()) {
				// rare case: allocate chunk
#if VERBOSE_ALLOC
				cerr << "new chunk of " << chunk_size
					<< " allocated." << endl;
#endif
				pool.push_back(chunk_type());
				chunk = &pool.back();
				chunk->reserve(chunk_size);
				INVARIANT_ASSERT(!chunk->size());
			}
			INVARIANT_ASSERT(chunk->capacity() == chunk_size);
			chunk->push_back(T());
			// T must be copy-constructible
			pointer ret = &chunk->back();
			INVARIANT_ASSERT(ret);
#if VERBOSE_ALLOC
			cerr << "alloc from pool @ " << ret << endl;
#endif
			return ret;
		}
		// lock will expire at end-of-scope
	}

	/// allocate n elements without construction, unimplemented
	pointer
	allocate(size_type n, list_vector_pool<void>::const_pointer hint = 0);

	/// deallocates without destruction
	// any safety checks for multiple deletions?
	// if so, use set<pointer> for the free_list
	void
	deallocate(pointer p) {
#if THREADED_ALLOC
		Lock got_the_mutex;
#endif
		assert(p);
#if VERBOSE_ALLOC
		cerr << "returned " << p << " to free-list." << endl;
#endif
		free_list.push(p);
		// lock will expire at end-of-scope
	}

	/// deallocates without destruction, unimplemented
	void
	deallocate(pointer p, size_type n);

	/**
		Initialize *p by val.  
		The new operator is given a placed address, 
		which doesn't actually reallocate.  
		The value is copy-constructed in-place.  
	 */
	void
	construct(pointer p, const T& val) {
		assert(p);
		new(p) T(val);
	}

	/// destroys *p without deallocating
	void
	destroy(pointer p) {
		assert(p);
		p->~T();
	}

	// other miscellaneous methods
	size_type max_size() const;

	// rebinding typedef

};	// end class list_vector_pool

//-----------------------------------------------------------------------------
#if THREADED_ALLOC
// static initializer of STL mutex element
template<class T, bool Threaded>
mutex_type
list_vector_pool<T, Threaded>::the_mutex = PTHREAD_MUTEX_INITIALIZER;

#endif

//- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
template <class T>
inline
bool
operator == (const list_vector_pool<T>&, const list_vector_pool<T>&) {
	return true;
}

//- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
template <class T>
inline
bool
operator != (const list_vector_pool<T>&, const list_vector_pool<T>&) {
	return false;
}

//=============================================================================
}	// end namespace memory


#undef	INVARIANT_ASSERT

#endif	//	__LIST_VECTOR_POOL_H__

