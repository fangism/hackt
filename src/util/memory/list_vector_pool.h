/**
	\file "list_vector_pool.h"
	Simple template container-based memory pool.  
	Basically allocates a large chunk at a time.  

	$Id: list_vector_pool.h,v 1.8.20.1 2005/02/19 06:56:50 fang Exp $
 */

#ifndef	__UTIL_MEMORY_LIST_VECTOR_POOL_H__
#define	__UTIL_MEMORY_LIST_VECTOR_POOL_H__

#include "memory/list_vector_pool_fwd.h"
#include "memory/thread_lock.h"
#include "macros.h"

#include <queue>
#include <list>
#include <vector>
#include <iosfwd>

// because FreeBSD's <pthreads.h> defines initializer as NULL
// and I haven't figured out a workaround yet.
#if defined(__FreeBSD__)
// #warn	"This allocator will not be thread-safe in FreeBSD... yet."
#define	THREADED_ALLOC		0
#else
#define	THREADED_ALLOC		1
#endif


#define	LIST_VECTOR_POOL_TEMPLATE_SIGNATURE				\
template <class T, bool Threaded>


//=============================================================================
// interface macros

/**
	\param T the class type to custom-allocate.
	\param C the chunk size to allocate for this class.  

        Non-member static allocation pool for a particular class.
        Note static linkage.  
        This static initialization must appear before any uses in methods.  

	Placement construct, needed by vector allocation.  
	Just No-op.  Does nothing.  

	Pool deallocation for this class.  
	A reinterpret_cast is needed because this particular allocator is 
	works with type-specific arguments, but the standard allocator 
	interface works with generic void-pointers.  
 */
#define	LIST_VECTOR_POOL_DEFAULT_STATIC_DEFINITION(T,C)			\
T::pool_type T::pool(C);						\
void* T::operator new (size_t s)					\
	{ return pool.allocate(); }					\
void* T::operator new (size_t s, void*& p)				\
	{ NEVER_NULL(p); return p; }					\
void T::operator delete (void* p)					\
	{ T* t = reinterpret_cast<T*>(p); NEVER_NULL(t); pool.deallocate(t); }

/**
	Convenient macro for explicitly requiring that a memory pool
	be ready during static initialization of a particular module.
	This is not required now... kept in comments for historical reference.
	\param T cannot be a template-id, use a convenient simple typedef.
 */
#define REQUIRES_LIST_VECTOR_POOL_STATIC_INIT(T)			\
static const T::pool_ref_type						\
__pool_ref_ ## T ## __ (T::get_pool());

/**
	These definitions are intended for using a reference-counted
	memory pool, as required by static global initialization ordering.  
	The static reference count object for the global pool will be 
	set by the call to acquire_pool_reference(), which is the only
	public interface to the allocator.  

	\param the name of the type.
	\param C the chunk size.  

	We cannot use a static count_ptr because we cannot guarantee
	that it will be initialized once by acquire_pool_reference()
	across all modules -- the home module may come along later and
	clobber it to NULL, because of object initialization.  
	Thus we must resort to plain-old-data (POD) 
	pool and pool_ref_count, which are guaranteed to be 
	NULL upon initialization.  

	Old obsolete notes on abandoning a reference-count scheme.  
	Fortunately, count_ptr gives us a means of faking reference 
	counts for such situations.  We return corecively constructed
	reference-count pointers with explicit count* arguments.  
	This didn't seem to destroy the last reference to the pool...
 */
#define	LIST_VECTOR_POOL_ROBUST_STATIC_DEFINITION(T,C)			\
									\
REQUIRES_LIST_VECTOR_POOL_STATIC_INIT(T)				\
									\
T::pool_ref_ref_type							\
T::get_pool(void) {							\
	static pool_type*	pool = new pool_type(C);		\
	static size_t*		count = new size_t(0);			\
	return pool_ref_ref_type(pool, count);				\
}									\
									\
void*									\
T::operator new (size_t s) {						\
	static pool_type& pool(*get_pool());				\
	LIST_VECTOR_POOL_STACKTRACE("operator new");			\
	return pool.allocate();						\
}									\
									\
inline									\
void*									\
T::operator new (size_t s, void*& p) {					\
	NEVER_NULL(p); return p;					\
}									\
									\
void									\
T::operator delete (void* p) {						\
	static pool_type& pool(*get_pool());				\
	LIST_VECTOR_POOL_STACKTRACE("operator delete");			\
	T* t = reinterpret_cast<T*>(p); NEVER_NULL(t);			\
	pool.deallocate(t);						\
}

									

//=============================================================================

namespace util {
//=============================================================================
/**
	Memory-management utility namespace.
 */
namespace memory {
using std::queue;
using std::list;
using std::vector;
using std::ostream;

//=============================================================================
/**
	Tag to dictate when destruction occurs in pool-managed resources.  
 */
struct lazy_destruction_tag { };

/**
	Tag to dictate when destruction occurs in pool-managed resources.  
 */
struct eager_destruction_tag { };

/**
	Default to eager / early destruction.  
	Is safe to use lazy destruction for non-recursive 
	destructors, i.e. object that do not contain pointers
	to other likewise pooled objects.  
	To override the default eager destruction, 
	define a specialization of this (in the util::memory namespace)
	in the module where the pool is instantiated.  
	A macro is provided below.  
	\param T the class type to be pooled.  
 */
template <class T>
struct list_vector_pool_policy {
	typedef	eager_destruction_tag	destruction_policy;
};	// end struct list_vector_pool_policy

/**
	This macro must appear in the util::memory namespace.  
 */
#define	LIST_VECTOR_POOL_LAZY_DESTRUCTION(T)				\
template <>								\
struct list_vector_pool_policy<T> {					\
	typedef	lazy_destruction_tag	destruction_policy;		\
};	// end struct list_vector_pool_policy

/**
	Helper method for distinguishing destruction policies.
 */
template <class T>
inline
typename list_vector_pool_policy<T>::destruction_policy
list_vector_pool_destruction_policy(void) {
	return typename list_vector_pool_policy<T>::destruction_policy();
}

//=============================================================================
// specialization
template <>
class list_vector_pool<void> {
public:
	typedef size_t		size_type;
	typedef ptrdiff_t	difference_type;
	typedef void*		pointer;
	typedef const void*	const_pointer;
	typedef void		value_type;
// no rebind yet...
};

//-----------------------------------------------------------------------------
#if 0
// Partial specialization for when allocated type is just pointer, 
// to prevent pointer recursion, not needed unless free_list_type is fancy
template <class T, bool Threaded>
class list_vector_pool<T*, Threaded> {
public:
	typedef	T*				value_type;
	typedef	size_t				size_type;
	typedef	ptrdiff_t			difference_type;
	typedef	value_type*			pointer;
	typedef	const value_type*		const_pointer;
	typedef	value_type&			reference;
	typedef	const value_type&		const_reference;
private:
	typedef	vector<value_type>		chunk_type;
	typedef	list<chunk_type>		impl_type;
	// plain queue to stop pointer recursion
	typedef	queue<pointer>			free_list_type;
public:

};	// end class list_vector_pool specialization
#endif

//-----------------------------------------------------------------------------
/**
	Simple, efficient memory pool.  
	No alignment or placement.  
	Should implement the std::allocator interface.  
	Only good for allocating one element at a time, 
	so is not general-purpose, not intended for arrays.  
	For a more general pool allocator use the __gnu_cxx::__pool_alloc, 
	which is likely to be the default allocator.  

	For stronger diagnostics.
	IDEA: use a bit-vector map to track which addresses are available.
	IDEA: use a discrete_interval_set to track address ranges.
 */
LIST_VECTOR_POOL_TEMPLATE_SIGNATURE
class list_vector_pool {
public:
	typedef	T				value_type;
	typedef	size_t				size_type;
	typedef	ptrdiff_t			difference_type;
	typedef	T*				pointer;
	typedef	const T*			const_pointer;
	typedef	T&				reference;
	typedef	const T&			const_reference;
private:
	// could propagate thread-sensitivity parameter to the respective
	// STL data-structures' allocators...
	typedef	vector<value_type>		chunk_type;
	typedef	list<chunk_type>		impl_type;
	typedef	pool_thread_lock<Threaded>	lock_type;
#if 1
	// could also try priority_queue... using less<pointer>
	// could try set<pointer> for uniqueness checking
		// set<> maintains begin() and end() with constant-time
	typedef	queue<pointer>			free_list_type;
#else
	// ballsy: use itself as allocator! (template recursion problem?)
	typedef	queue<pointer, deque<pointer, list_vector_pool<pointer> > >
						free_list_type;
#endif

	// probably no need to change...
	const size_t				chunk_size;
	impl_type				pool;
	free_list_type				free_list;
	/**
		Records the peak number of elements allocated.
		Useful for detecting memory leaks.  
	 */
	size_t					peak;

	// mutex shouldn't be static!  should be PER pool
	mutex_type				the_mutex;

public:
	explicit
	list_vector_pool(const size_type C = 16);

	// no copy-constructor

	~list_vector_pool();

	pointer
	allocate(void);

	/// allocate n elements without construction, unimplemented
	pointer
	allocate(size_type n, list_vector_pool<void>::const_pointer hint = 0);

	void
	deallocate(pointer p);

	/// deallocates, unimplemented
	void
	deallocate(pointer p, size_type n);

	void
	construct(pointer p, const T& val);

	void
	destroy(pointer p);

	// other miscellaneous methods
	size_type
	max_size() const;

	// rebinding typedef

	/// feedback IO, indented one-tab by default
	ostream&
	status(ostream& o) const;

private:
	// policy-based variations

	static
	void
	eager_destroy(const pointer, const eager_destruction_tag);

	static
	void
	eager_destroy(const pointer, const lazy_destruction_tag);

	static
	void
	lazy_destroy(const pointer, const eager_destruction_tag);

	static
	void
	lazy_destroy(const pointer, const lazy_destruction_tag);

};	// end class list_vector_pool

//-----------------------------------------------------------------------------
template <class T>
bool
operator == (const list_vector_pool<T>&, const list_vector_pool<T>&);

//- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
template <class T>
bool
operator != (const list_vector_pool<T>&, const list_vector_pool<T>&);

//=============================================================================
}	// end namespace memory
}	// end namespace util

#endif	// __UTIL_MEMORY_LIST_VECTOR_POOL_H__

