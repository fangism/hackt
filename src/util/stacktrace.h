/**
	\file "stacktrace.h"
	Utility macros and header for convenient stack-trace debugging.
	$Id: stacktrace.h,v 1.5 2005/01/15 06:20:46 fang Exp $
 */

#ifndef	__STACK_TRACE_H__
#define	__STACK_TRACE_H__

// macros for enabling/disabling stacktrace code
/**
	Strongly recommend using these macros to be able to 
	turn everything off at compile time.  
	Predefine this to 0 at compile time to turn-off.  
 */
#ifndef	ENABLE_STACKTRACE
#define	ENABLE_STACKTRACE	0	// on (1) or off (0) by default
#endif


// This is the macro interface intended for the programmer.  
#if ENABLE_STACKTRACE
	#define	USING_STACKTRACE	using util::stacktrace;
	#define	STACKTRACE(str)	stacktrace __stacktrace__(str)
	/**
		This enables echoing each time trace stack is updated, i.e., 
		upon entering and leaving function call stack 
		or lexical scopes.  
	 */
	#define STACKTRACE_ECHO_ON					\
			stacktrace::echo __echo_stacktrace__(1)
	#define STACKTRACE_ECHO_OFF					\
			stacktrace::echo __echo_stacktrace__(0)
	#define	STACKTRACE_STREAM					\
			stacktrace::stream()
	#define REDIRECT_STACKTRACE(os)					\
			stacktrace::redirect __redir_stacktrace__(os)
	#define	ASSERT_STACKTRACE(expr)					\
			if (!(expr)) { stacktrace::full_dump(); assert(expr); }
#else
	#define	USING_STACKTRACE
	#define	STACKTRACE(str)
	#define STACKTRACE_ECHO_ON
	#define STACKTRACE_ECHO_OFF
	#define	STACKTRACE_STREAM		std::cerr
	#define REDIRECT_STACKTRACE(os)
	#define	ASSERT_STACKTRACE(expr)		assert(expr)
#endif


//=============================================================================

#if ENABLE_STACKTRACE

#include <iosfwd>
#include "macros.h"
#include "STL/list_fwd.h"
#include "string_fwd.h"


namespace util {
USING_LIST
using std::ostream;
using std::string;

//=============================================================================
/**
	Nothing is inlined because when you're debugging you shouldn't be
	varing about performance.  
	Can use for function scopes or any conditional statement scopes.  
	Two mechanisms for controlling behavior, 
	compile-time macros and public static variables.  
	TODO: make one entry per thread, mapped by thread ID.
 */
class stacktrace {
public:
	class manager;
	struct echo;
	struct redirect;

public:
	stacktrace(const string&);
	~stacktrace();
public:
	static
	ostream&
	stream(void);

	/**
		Explicit request by user to dump the stack trace.
		Useful in assertion failures.  
	 */
	static
	void
	full_dump(void);
};	// end class stacktrace

//-----------------------------------------------------------------------------
/**
	Whether or not to print upon entering and exiting.
	Pass in 0 to disable.  
	Enabling/disable lasts for the duration of the scope.  
 */
struct stacktrace::echo {
	echo(const int i = 1);
	~echo();
};

//-----------------------------------------------------------------------------
/**
	Redirect all stack dumps to this ostream until changed otherwise.
	Lasts for the duration of the scope where this is called.  
 */
struct stacktrace::redirect {
	redirect(ostream&);
	~redirect();
};

//=============================================================================

}	// end namespace util

#else
	// don't even bother processing class declaration!
#endif	// ENABLE_STACKTRACE

#endif	// __STACK_TRACE_H__

