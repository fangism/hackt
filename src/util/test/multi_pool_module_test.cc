/**
	\file "multi_pool_module_test.cc"
	Testing robust static initialization of separate memory pools.
	This is just a dummy file to provide a main symbol.
	$Id: multi_pool_module_test.cc,v 1.6 2006/02/26 02:28:04 fang Exp $
 */

#define	ENABLE_STACKTRACE		1
#define	ENABLE_STATIC_TRACE		1

#ifdef	NDEBUG
#undef	NDEBUG
#endif

#include <iostream>
#include "util/using_ostream.hh"

#include "pool_module_a.hh"
#include "pool_module_b.hh"

#include "util/memory/list_vector_pool.hh"
#include "util/stacktrace.hh"
#include "util/static_trace.hh"

STATIC_TRACE_BEGIN("test-main")

// not required because we have no static globals
// REQUIRES_LIST_VECTOR_POOL_STATIC_INIT(twiddle_dum)
// REQUIRES_LIST_VECTOR_POOL_STATIC_INIT(twiddle_dee)

static void
never_let_go(void) {
	STACKTRACE("never_let_go()");
	static excl_ptr<twiddle_dum> hold_on1(new twiddle_dum);
	static excl_ptr<twiddle_dee> hold_on2(new twiddle_dee);
}

int
main(int, char*[]) {
	STACKTRACE("MAIN()");	// magic string used for filtering for test
	cerr << "done with static initialization, start of main()" << endl;
#if 1
	never_let_go();
	never_let_go();
	never_let_go();
#endif
	cerr << "done with main(), begin static destruction." << endl;
	return 0;
}

STATIC_TRACE_END("test-main")

