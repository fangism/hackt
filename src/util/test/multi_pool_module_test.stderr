START-initialization: twiddle_dee module
\-{ list_vector_pool()
Reserved twiddle_dum chunk of size 8*1 starting at 0x8061260
/-} list_vector_pool()
\-{ list_vector_pool()
Reserved twiddle_dee chunk of size 8*1 starting at 0x8061308
/-} list_vector_pool()
\-{ operator new
| \-{ list_vector_pool::allocate()
| | \-{ construct twiddle_dum
at 0xbfcc9c90
| | /-} construct twiddle_dum
| | \-{ destroy twiddle_dum
at 0xbfcc9c90
| | /-} destroy twiddle_dum
Allocated twiddle_dum from pool @ 0x8061260
| /-} list_vector_pool::allocate()
/-} operator new
\-{ construct twiddle_dum
at 0x8061260
/-} construct twiddle_dum
\-{ operator new
| \-{ list_vector_pool::allocate()
| | \-{ construct twiddle_dum
at 0xbfcc9c90
| | /-} construct twiddle_dum
| | \-{ destroy twiddle_dum
at 0xbfcc9c90
| | /-} destroy twiddle_dum
Allocated twiddle_dum from pool @ 0x8061261
| /-} list_vector_pool::allocate()
/-} operator new
\-{ construct twiddle_dum
at 0x8061261
/-} construct twiddle_dum
\-{ operator new
| \-{ list_vector_pool::allocate()
| | \-{ construct twiddle_dee
at 0xbfcc9c90
| | /-} construct twiddle_dee
| | \-{ destroy twiddle_dee
at 0xbfcc9c90
| | /-} destroy twiddle_dee
Allocated twiddle_dee from pool @ 0x8061308
| /-} list_vector_pool::allocate()
/-} operator new
\-{ construct twiddle_dee
at 0x8061308
/-} construct twiddle_dee
\-{ operator new
| \-{ list_vector_pool::allocate()
| | \-{ construct twiddle_dee
at 0xbfcc9c90
| | /-} construct twiddle_dee
| | \-{ destroy twiddle_dee
at 0xbfcc9c90
| | /-} destroy twiddle_dee
Allocated twiddle_dee from pool @ 0x8061309
| /-} list_vector_pool::allocate()
/-} operator new
\-{ construct twiddle_dee
at 0x8061309
/-} construct twiddle_dee
END---initialization: twiddle_dee module
START-initialization: twiddle_dum module
\-{ operator new
| \-{ list_vector_pool::allocate()
| | \-{ construct twiddle_dee
at 0xbfcc9c90
| | /-} construct twiddle_dee
| | \-{ destroy twiddle_dee
at 0xbfcc9c90
| | /-} destroy twiddle_dee
Allocated twiddle_dee from pool @ 0x806130a
| /-} list_vector_pool::allocate()
/-} operator new
\-{ construct twiddle_dee
at 0x806130a
/-} construct twiddle_dee
\-{ operator new
| \-{ list_vector_pool::allocate()
| | \-{ construct twiddle_dee
at 0xbfcc9c90
| | /-} construct twiddle_dee
| | \-{ destroy twiddle_dee
at 0xbfcc9c90
| | /-} destroy twiddle_dee
Allocated twiddle_dee from pool @ 0x806130b
| /-} list_vector_pool::allocate()
/-} operator new
\-{ construct twiddle_dee
at 0x806130b
/-} construct twiddle_dee
\-{ operator new
| \-{ list_vector_pool::allocate()
| | \-{ construct twiddle_dum
at 0xbfcc9c90
| | /-} construct twiddle_dum
| | \-{ destroy twiddle_dum
at 0xbfcc9c90
| | /-} destroy twiddle_dum
Allocated twiddle_dum from pool @ 0x8061262
| /-} list_vector_pool::allocate()
/-} operator new
\-{ construct twiddle_dum
at 0x8061262
/-} construct twiddle_dum
\-{ operator new
| \-{ list_vector_pool::allocate()
| | \-{ construct twiddle_dum
at 0xbfcc9c90
| | /-} construct twiddle_dum
| | \-{ destroy twiddle_dum
at 0xbfcc9c90
| | /-} destroy twiddle_dum
Allocated twiddle_dum from pool @ 0x8061263
| /-} list_vector_pool::allocate()
/-} operator new
\-{ construct twiddle_dum
at 0x8061263
/-} construct twiddle_dum
END---initialization: twiddle_dum module
START-initialization: test-main
END---initialization: test-main
\-{ MAIN()
done with static initialization, start of main()
| \-{ never_let_go()
| | \-{ operator new
| | | \-{ list_vector_pool::allocate()
| | | | \-{ construct twiddle_dum
at 0xbfcc9c80
| | | | /-} construct twiddle_dum
| | | | \-{ destroy twiddle_dum
at 0xbfcc9c80
| | | | /-} destroy twiddle_dum
Allocated twiddle_dum from pool @ 0x8061264
| | | /-} list_vector_pool::allocate()
| | /-} operator new
| | \-{ construct twiddle_dum
at 0x8061264
| | /-} construct twiddle_dum
| | \-{ operator new
| | | \-{ list_vector_pool::allocate()
| | | | \-{ construct twiddle_dee
at 0xbfcc9c80
| | | | /-} construct twiddle_dee
| | | | \-{ destroy twiddle_dee
at 0xbfcc9c80
| | | | /-} destroy twiddle_dee
Allocated twiddle_dee from pool @ 0x806130c
| | | /-} list_vector_pool::allocate()
| | /-} operator new
| | \-{ construct twiddle_dee
at 0x806130c
| | /-} construct twiddle_dee
| /-} never_let_go()
| \-{ never_let_go()
| /-} never_let_go()
| \-{ never_let_go()
| /-} never_let_go()
done with main(), begin static destruction.
/-} MAIN()
\-{ destroy twiddle_dee
at 0x806130c
/-} destroy twiddle_dee
\-{ operator delete
| \-{ list_vector_pool::deallocate()
Returned twiddle_dee @ 0x806130c to free-list.
| /-} list_vector_pool::deallocate()
/-} operator delete
\-{ destroy twiddle_dum
at 0x8061264
/-} destroy twiddle_dum
\-{ operator delete
| \-{ list_vector_pool::deallocate()
Returned twiddle_dum @ 0x8061264 to free-list.
| /-} list_vector_pool::deallocate()
/-} operator delete
START-destruction: test-main
END---destruction: test-main
START-destruction: twiddle_dum module
\-{ destroy twiddle_dum
at 0x8061263
/-} destroy twiddle_dum
\-{ operator delete
| \-{ list_vector_pool::deallocate()
Returned twiddle_dum @ 0x8061263 to free-list.
| /-} list_vector_pool::deallocate()
/-} operator delete
\-{ destroy twiddle_dum
at 0x8061262
/-} destroy twiddle_dum
\-{ operator delete
| \-{ list_vector_pool::deallocate()
Returned twiddle_dum @ 0x8061262 to free-list.
| /-} list_vector_pool::deallocate()
/-} operator delete
\-{ destroy twiddle_dee
at 0x806130b
/-} destroy twiddle_dee
\-{ operator delete
| \-{ list_vector_pool::deallocate()
Returned twiddle_dee @ 0x806130b to free-list.
| /-} list_vector_pool::deallocate()
/-} operator delete
\-{ destroy twiddle_dee
at 0x806130a
/-} destroy twiddle_dee
\-{ operator delete
| \-{ list_vector_pool::deallocate()
Returned twiddle_dee @ 0x806130a to free-list.
| /-} list_vector_pool::deallocate()
/-} operator delete
END---destruction: twiddle_dum module
START-destruction: twiddle_dee module
\-{ destroy twiddle_dee
at 0x8061309
/-} destroy twiddle_dee
\-{ operator delete
| \-{ list_vector_pool::deallocate()
Returned twiddle_dee @ 0x8061309 to free-list.
| /-} list_vector_pool::deallocate()
/-} operator delete
\-{ destroy twiddle_dee
at 0x8061308
/-} destroy twiddle_dee
\-{ operator delete
| \-{ list_vector_pool::deallocate()
Returned twiddle_dee @ 0x8061308 to free-list.
| /-} list_vector_pool::deallocate()
/-} operator delete
\-{ destroy twiddle_dum
at 0x8061261
/-} destroy twiddle_dum
\-{ operator delete
| \-{ list_vector_pool::deallocate()
Returned twiddle_dum @ 0x8061261 to free-list.
| /-} list_vector_pool::deallocate()
/-} operator delete
\-{ destroy twiddle_dum
at 0x8061260
/-} destroy twiddle_dum
\-{ operator delete
| \-{ list_vector_pool::deallocate()
Returned twiddle_dum @ 0x8061260 to free-list.
| /-} list_vector_pool::deallocate()
/-} operator delete
\-{ ~list_vector_pool()
~list_vector_pool<twiddle_dee>() at 0x8061270
| \-{ list_vector_pool::status()
	1 chunks of 8*1 twiddle_dee allocated.
	Peak usage: 5 elements, free-list has 5 remaining.
| /-} list_vector_pool::status()
| \-{ construct twiddle_dee
at 0x806130c
| /-} construct twiddle_dee
| \-{ construct twiddle_dee
at 0x806130b
| /-} construct twiddle_dee
| \-{ construct twiddle_dee
at 0x806130a
| /-} construct twiddle_dee
| \-{ construct twiddle_dee
at 0x8061309
| /-} construct twiddle_dee
| \-{ construct twiddle_dee
at 0x8061308
| /-} construct twiddle_dee
/-} ~list_vector_pool()
\-{ destroy twiddle_dee
at 0x8061308
/-} destroy twiddle_dee
\-{ destroy twiddle_dee
at 0x8061309
/-} destroy twiddle_dee
\-{ destroy twiddle_dee
at 0x806130a
/-} destroy twiddle_dee
\-{ destroy twiddle_dee
at 0x806130b
/-} destroy twiddle_dee
\-{ destroy twiddle_dee
at 0x806130c
/-} destroy twiddle_dee
\-{ ~list_vector_pool()
~list_vector_pool<twiddle_dum>() at 0x8061148
| \-{ list_vector_pool::status()
	1 chunks of 8*1 twiddle_dum allocated.
	Peak usage: 5 elements, free-list has 5 remaining.
| /-} list_vector_pool::status()
/-} ~list_vector_pool()
\-{ destroy twiddle_dum
at 0x8061260
/-} destroy twiddle_dum
\-{ destroy twiddle_dum
at 0x8061261
/-} destroy twiddle_dum
\-{ destroy twiddle_dum
at 0x8061262
/-} destroy twiddle_dum
\-{ destroy twiddle_dum
at 0x8061263
/-} destroy twiddle_dum
\-{ destroy twiddle_dum
at 0x8061264
/-} destroy twiddle_dum
END---destruction: twiddle_dee module
